Source/Particles/AddPlasmaUtilities.H

```diff
@@ -71,12 +94,110 @@ int compute_area_weights (const amrex::IntVect& iv, const int normal_axis) {
+amrex::Real compute_scale_fac_area_eb (
+    const amrex::GpuArray<amrex::Real, AMREX_SPACEDIM>& dx,
+    const amrex::Real num_ppc_real,
+    amrex::Array4<const amrex::Real> const& eb_bnd_normal_arr,
+    int i, int j, int k ) {
+    using namespace amrex::literals;
+    // Scale particle weight by the area of the emitting surface, within one cell
+    // By definition, eb_bnd_area_arr is normalized (unitless).
+    // Here we undo the normalization (i.e. multiply by the surface used for normalization in amrex:
+    // see https://amrex-codes.github.io/amrex/docs_html/EB.html#embedded-boundary-data-structures)
+#if defined(WARPX_DIM_3D)
+    const amrex::Real nx = eb_bnd_normal_arr(i,j,k,0);
+    const amrex::Real ny = eb_bnd_normal_arr(i,j,k,1);
+    const amrex::Real nz = eb_bnd_normal_arr(i,j,k,2);
+    amrex::Real scale_fac = std::sqrt(amrex::Math::powi<2>(nx*dx[1]*dx[2]) +
+                  amrex::Math::powi<2>(ny*dx[0]*dx[2]) +
+                  amrex::Math::powi<2>(nz*dx[0]*dx[1]));
+
+#elif defined(WARPX_DIM_RZ) || defined(WARPX_DIM_XZ)
+    const amrex::Real nx = eb_bnd_normal_arr(i,j,k,0);
+    const amrex::Real nz = eb_bnd_normal_arr(i,j,k,1);
+    amrex::Real scale_fac = std::sqrt(amrex::Math::powi<2>(nx*dx[1]) +
+                  amrex::Math::powi<2>(nz*dx[0]));
+#else
+    amrex::ignore_unused(dx, eb_bnd_normal_arr, i, j, k);
+    amrex::Real scale_fac = 0.0_rt;
+#endif
+    // Do not multiply by eb_bnd_area_arr(i,j,k) here because this
+    // already taken into account by emitting a number of macroparticles
+    // that is proportional to the area of eb_bnd_area_arr(i,j,k).
+    scale_fac /= num_ppc_real;
+    return scale_fac;
+}
+
+/* Rotate the momentum of the particle, using a rotation that transforms the
+   z axis to the normal of the embedded boundary. More specifically, before
+   calling this function, `pu.z` has the momentum distribution that is
+   meant for the direction normal to the embedded boundary, and `pu.x`/`pu.y`
+   have the momentum distribution that is meant for the tangentional direction.
+   After calling this function, `pu.x`, `pu.y`, `pu.z` will have the correct
+   momentum distribution, consistent with the local normal to the embedded boundary. */
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+void rotate_momentum_eb (
+    PDim3 & pu,
+    amrex::Array4<const amrex::Real> const& eb_bnd_normal_arr,
+    int i, int j, int k )
+{
+#if defined(WARPX_DIM_3D)
+    // The minus sign below takes into account the fact that eb_bnd_normal_arr
+    // points towards the covered region, while particles are to be emitted
+    // *away* from the covered region.
+    amrex::Real const nx = -eb_bnd_normal_arr(i,j,k,0);
+    amrex::Real const ny = -eb_bnd_normal_arr(i,j,k,1);
+    amrex::Real const nz = -eb_bnd_normal_arr(i,j,k,2);
+
+    // Rotate the momentum along the theta direction (in the z-x plane)
+    amrex::Real const cos_theta = nz;
+    amrex::Real const sin_theta = std::sqrt(1-nz*nz);
```

Comment:
1._rt

To simplify this, could the theta and phi rotations be combined into one set of expressions, a 3D rotation?